### ISSUE TITLE
error: [Errno 24] Too many open files

### ISSUE THREAD

**Issue by zacharyblank** [mitmproxy/mitmproxy#383]
When running `mitmdump -s /var/monitor/log.py -T --host`

```
<< 502: Connection closed prematurely
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/local/lib/python2.7/dist-packages/libmproxy/controller.py", line 77, in run
  File "/usr/local/lib/python2.7/dist-packages/netlib/tcp.py", line 397, in serve_forever
  File "/usr/lib/python2.7/socket.py", line 202, in accept
error: [Errno 24] Too many open files
```

**Comment by limeric29** [comment]
Hello,

You received a gift of 7-day VPN service. You can now surf the Internet anonymously and even visit blocked websites and mobile apps.

http://hideme.io

**Comment by thomir** [comment]
Hi,

I'm seeing something similar:

```
Exception in thread SlaveThread (0.0.0.0:8080):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/home/thomi/code/mitmproxy/local/lib/python2.7/site-packages/libmproxy/controller.py", line 77, in run
    self.server.serve_forever()
  File "/home/thomi/code/mitmproxy/local/lib/python2.7/site-packages/netlib/tcp.py", line 479, in serve_forever
    connection, client_address = self.socket.accept()
  File "/usr/lib/python2.7/socket.py", line 202, in accept
    sock, addr = self._sock.accept()
error: [Errno 24] Too many open files
```

I'm running mitmproxy 0.11.3. 

I get this error the very first time mitmproxy receives any traffic. The output of 'lsof -n -p <mitmproxy_pid>' is:
```
COMMAND     PID  USER   FD   TYPE  DEVICE SIZE/OFF     NODE NAME
mitmproxy 32258 thomi  cwd    DIR     8,1     4096  9321880 /home/thomi/code/mitmproxy
mitmproxy 32258 thomi  rtd    DIR     8,1     4096        2 /
mitmproxy 32258 thomi  txt    REG     8,1  3781768 10098076 /home/thomi/code/mitmproxy/bin/python2
mitmproxy 32258 thomi  mem    REG     8,1    18472 10747941 /lib/x86_64-linux-gnu/libgpg-error.so.0.11.0
mitmproxy 32258 thomi  mem    REG     8,1   137400 10748133 /lib/x86_64-linux-gnu/liblzma.so.5.0.0
mitmproxy 32258 thomi  mem    REG     8,1   526144 10748369 /lib/x86_64-linux-gnu/libgcrypt.so.11.8.3
mitmproxy 32258 thomi  mem    REG     8,1  1461720  6160494 /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1
mitmproxy 32258 thomi  mem    REG     8,1    83816  6177395 /usr/lib/x86_64-linux-gnu/libexslt.so.0.8.17
mitmproxy 32258 thomi  mem    REG     8,1   247952  6177394 /usr/lib/x86_64-linux-gnu/libxslt.so.1.1.28
mitmproxy 32258 thomi  mem    REG     8,1  8414048 10235332 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/lxml/etree.so
mitmproxy 32258 thomi  mem    REG     8,1   281288  6168844 /usr/lib/x86_64-linux-gnu/libjpeg.so.8.0.2
mitmproxy 32258 thomi  mem    REG     8,1  1229080 10235928 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/PIL/_imaging.so
mitmproxy 32258 thomi  mem    REG     8,1    34904  6309953 /usr/lib/python2.7/lib-dynload/_csv.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1   167128 10748241 /lib/x86_64-linux-gnu/libtinfo.so.5.9
mitmproxy 32258 thomi  mem    REG     8,1   211784 10748351 /lib/x86_64-linux-gnu/libncursesw.so.5.9
mitmproxy 32258 thomi  mem    REG     8,1    86584  6309954 /usr/lib/python2.7/lib-dynload/_curses.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1    17640 10234681 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/tornado/speedups.so
mitmproxy 32258 thomi  mem    REG     8,1    24872  6309938 /usr/lib/python2.7/lib-dynload/termios.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1    46448 10232933 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/urwid/str_util.so
mitmproxy 32258 thomi  mem    REG     8,1    54248  6308903 /usr/lib/python2.7/lib-dynload/_json.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1  2480344 10236499 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/cryptography/_Cryptography_cffi_4ee21876x41fb9936.so
mitmproxy 32258 thomi  mem    REG     8,1   272488 10102402 /home/thomi/code/mitmproxy/lib/python2.7/site-packages/_cffi_backend.so
mitmproxy 32258 thomi  mem    REG     8,1   387432 10748605 /lib/x86_64-linux-gnu/libssl.so.1.0.0
mitmproxy 32258 thomi  mem    REG     8,1    97872  6309948 /usr/lib/python2.7/lib-dynload/_ssl.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1    19000 10748548 /lib/x86_64-linux-gnu/libuuid.so.1.3.0
mitmproxy 32258 thomi  mem    REG     8,1  1961344 10748234 /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
mitmproxy 32258 thomi  mem    REG     8,1    29464  6309958 /usr/lib/python2.7/lib-dynload/_hashlib.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1    30944  6161032 /usr/lib/x86_64-linux-gnu/libffi.so.6.0.2
mitmproxy 32258 thomi  mem    REG     8,1   141184  6309945 /usr/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so
mitmproxy 32258 thomi  mem    REG     8,1  5680496  6160673 /usr/lib/locale/locale-archive
mitmproxy 32258 thomi  mem    REG     8,1  1076720 10748258 /lib/x86_64-linux-gnu/libm-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1   100728 10751980 /lib/x86_64-linux-gnu/libz.so.1.2.8
mitmproxy 32258 thomi  mem    REG     8,1    10680 10748226 /lib/x86_64-linux-gnu/libutil-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1    14664 10748142 /lib/x86_64-linux-gnu/libdl-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1  1844160 10748275 /lib/x86_64-linux-gnu/libc-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1   145224 10748286 /lib/x86_64-linux-gnu/libpthread-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1   150552 10748138 /lib/x86_64-linux-gnu/ld-2.19.so
mitmproxy 32258 thomi  mem    REG     8,1    26258  6563203 /usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache
mitmproxy 32258 thomi    0u   CHR  136,40      0t0       43 /dev/pts/40
mitmproxy 32258 thomi    1u   CHR  136,40      0t0       43 /dev/pts/40
mitmproxy 32258 thomi    2u   CHR  136,40      0t0       43 /dev/pts/40
mitmproxy 32258 thomi    3u  IPv4 3196046      0t0      TCP *:http-alt (LISTEN)
mitmproxy 32258 thomi    4r  FIFO     0,8      0t0  3196047 pipe
mitmproxy 32258 thomi    5w  FIFO     0,8      0t0  3196047 pipe
mitmproxy 32258 thomi   12r   CHR     1,9      0t0     1034 /dev/urandom
```

As this is the first time I'm using mitmproxy in anger, I suspect this is due to a misconfiguration on my part, but even so, it'd be nice if mitmproxy was able to tell me what I've done wrong, or at least not crash :D

**Comment by andresriancho** [comment]
I'm also hitting this bug, seems to be pretty common

**Comment by mhils** [comment]
@andresriancho, thanks for the report you linked to. Can you reproduce this deterministically? Which OS?

Thanks,
Max

**Comment by andresriancho** [comment]
Yup, I can reproduce it each time.

Console 1:
```
mitmproxy -T -p 8888
```

Console 2:
```
nosetests -s w3af/plugins/attack/db/tests/test_sqlmap_wrapper.py:TestSQLMapWrapper.test_verify_vulnerability_ssl
```

Console 1:
```
Exception in thread SlaveThread (0.0.0.0:8888):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/home/pablo/pch/virtual-envs/w3af/local/lib/python2.7/site-packages/libmproxy/controller.py", line 77, in run
  File "/home/pablo/pch/virtual-envs/w3af/local/lib/python2.7/site-packages/netlib/tcp.py", line 479, in serve_forever
  File "/usr/lib/python2.7/socket.py", line 202, in accept
error: [Errno 24] Too many open files
```

To be able to run this you'll have to download w3af (develop branch) and install all the dependencies.

I'm using ubuntu 14.04:
```
$ ulimit -n -v
open files                      (-n) 1024
```

PS: Been looking around and it seems that this is one of the ugliest bugs around, hopefully it will get fixed quickly :+1:

**Comment by andresriancho** [comment]
Tested with both the latest from `pypi` and from the repo's `master`. To test the latest from master I run:

```
pip install 'git+https://github.com/mitmproxy/mitmproxy.git' --upgrade
pip install --upgrade 'git+https://github.com/mitmproxy/netlib.git'
```

**Comment by mhils** [comment]
@andresriancho, mitmproxy and w3af on the same machine?

**Comment by andresriancho** [comment]
yup, same box, just different consoles.

**Comment by andresriancho** [comment]
Ah, just remembered, in `test_sqlmap_wrapper.py` you'll have to apply this patch:

```patch
-self.local_proxy_url = 'http://%s:%s/' % (host, self.proxy.get_bind_port())
+self.local_proxy_url = 'http://%s:%s/' % (host, 8888)
```

**Comment by andresriancho** [comment]
If you've got any questions on how to repro, I'm at #w3af on freenode: `__apr__`

**Comment by Nerudifus** [comment]
This issue is still marked as "open" thus I use it instead of opening a new one.

I'm running mitmdump as transparent proxy with SSL. Some of the clients do not have the certificate installed and this is intended. After running some hours the clients could not open any website. At the end of the log/output there is
```
Exception in thread SlaveThread (0.0.0.0:8080):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/local/lib/python2.7/dist-packages/libmproxy/controller.py",
line 77, in run
  File "/usr/local/lib/python2.7/dist-packages/netlib/tcp.py", line 479,
in serve_forever
  File "/usr/lib/python2.7/socket.py", line 202, in accept
error: [Errno 24] Too many open files
```
Previous to this there are multiple ```ProxyError(NetLibError('Error connecting to
"[Destination IP]": [Errno 24] Too many open files',),)``` lines.
In the whole stdout there are lots (>2000) of
```
192.168.W.Q:47777: clientconnect
192.168.W.Q:47777: SSL handshake error: The client may not trust the proxy's certificate.
```
But the corresponding ```192.168.W.Q:47777: clientdisconnect ``` is missing. Thus it seems that the connection is opened but never closed. I am pretty sure that the clients do not keep the connection opened. They (soft) restart after some time.

All this was done with version 0.11.3. As far as I can tell none of the three options mentioned in the last post of 2015-02-03 was true.
Now I setup a clean new system with the current version 0.12.1 and tried to reproduce the error. Sadly I found no good way to reproduce the SSL handshake errors. 
In both cases mitmdump run under Ubuntu 14.04 with the newest updates at that time.

It was possible to get some (>1000) of them with one client but not enough to crash mitmdump. 
At the end ```lsof -n``` lists more than 1000 lines like ```mitmdump [...] sock [...] 0,7      0t0   [...] can't identify protocol``` and more than 4500 lines like ```mitmdump [...] IPv4 [...] 0t0  TCP [...] (ESTABLISHED) ```. This seems a lot but for sock there are only 18 unique lines repeating the block 56 times only with another TID. The same goes for the IPv4 lines, there are only 82 unique lines which repeats 56 times only with other TIDs.
In the output of mitmdump there are the lines for ```clientconnect``` and ```SSL handshake error``` but no disconnect. Additionally the port of the client used to connect is reused after some time. Thus the client thinks all connections to/with the port are closed.

The test method was to use one client with Windows 7 with all updates and Firefox 38.0.5 (without any proxy configure). Two different website were used. In an automatic way Firefox was started with one website and after some waiting time it was close. Then Firefox was started again with the other website and closed after some time. The complete process was repeated multiple times. I do not now if it was 56 times or more.
Nearly at the end the xserver or display manager of the mitmdump computer did not response any more. Even the switch to console was not displayed any more but done in background. The client still could open the websites at that time. After some waiting time the lsof was done and lightdm restarted. This took some time but afterwards everything works normal but mitmdump was closed.

All in all I could confirm that an ```[Errno 24] Too many open files``` error also occur in version 0.11.3 but it is not sure if this is true for 0.12.1, too. But the problem with not listed ```clientdisconnect``` is true for both versions. But maybe work as intended or is just a problem of the output.

This post is created in the hope that it may present some information. I will do more tests in the next days and may have more resp. better information then.
